pkg/virt-api/webhooks/arm64.go

```diff
@@ -0,0 +1,156 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2019 Red Hat, Inc.
+ *
+ */
+
+/*
+ * arm64 utilities are in the webhooks package because they are used both
+ * by validation and mutation webhooks.
+ */
+package webhooks
+
+import (
+        "fmt"
+	rt "runtime"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+
+        v1 "kubevirt.io/client-go/api/v1"
+)
+
+var _false bool = false
+
+const (
+	defaultCPUModel = "host-passthrough"
+)
+
+// verifyInvalidSetting verify if VMI spec contain unavailable setting for arm64, check following items:
+// 1. if setting bios boot
+// 2. if use uefi secure boot
+// 3. if set auto-attach graphics device
+// 4. if use host-model for cpu model
+func verifyInvalidSetting(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec) (metav1.StatusCause, bool) {
+	if spec.Domain.Firmware != nil && spec.Domain.Firmware.Bootloader !=nil {
+		if spec.Domain.Firmware.Bootloader.BIOS != nil {
+			return metav1.StatusCause{
+				Type:		metav1.CauseTypeFieldValueNotSupported,
+				Message:	"Arm64 does not support bios boot, please change to uefi boot",
+				Field:		field.Child("domain", "firmware", "bootloader", "bios").String(),
+			} , false
+		}
+		if spec.Domain.Firmware.Bootloader.EFI != nil {
+			// When EFI is enable, secureboot is enabled by default, so here check two condition
+			// 1 is EFI is enabled without Secureboot setting
+			// 2 is both EFI and Secureboot enabled
+			if spec.Domain.Firmware.Bootloader.EFI.SecureBoot == nil || (spec.Domain.Firmware.Bootloader.EFI.SecureBoot != nil && *spec.Domain.Firmware.Bootloader.EFI.SecureBoot) {
+				return metav1.StatusCause{
+					Type:		metav1.CauseTypeFieldValueNotSupported,
+					Message:	"UEFI secure boot is currently not supported on aarch64 Arch",
+					Field:		field.Child("domain", "firmware", "bootloader", "efi", "secureboot").String(),
+				} , false
+			}
+		}
+	}
+	if &spec.Domain.Devices != nil && spec.Domain.Devices.AutoattachGraphicsDevice != nil && *spec.Domain.Devices.AutoattachGraphicsDevice {
+		return metav1.StatusCause{
+			Type:		metav1.CauseTypeFieldValueNotSupported,
+			Message:	"VGA device is not support by buildin qemu-kvm in virt-launcher image for Arm64, please disable autoattachGraphicsDevice",
+			Field:		field.Child("domain", "device", "autoattachgraphicsdevice").String(),
+		} , false
+	}
+	if spec.Domain.CPU != nil && &spec.Domain.CPU.Model != nil && spec.Domain.CPU.Model == "host-model" {
+		return metav1.StatusCause{
+			Type:		metav1.CauseTypeFieldValueNotSupported,
+			Message:	"Arm64 not support host model well",
+			Field:		field.Child("domain", "cpu", "model").String(),
+		} , false
+	}
+	return metav1.StatusCause{}, true
+}
+
+// verifyArch verify if arch is arm64
+func verifyArch() bool {
+	if rt.GOARCH == "arm64" {
+		return true
+	}
+	return false
+}
```

Comment:
this webhook logic all makes sense to me. Can we get a unit test to exercise this logic?

Because of the way you've gated all the arm defaults/validation behind this `verifyArch()` function, it would actually be pretty trivial to mock this arm validation/defaulting for all architectures.

at the top of the file add something like this...
```
var VerifyArchFunc = defaultVerifyArchFunc
```
then rename verifyArch function to defaultVerifyArchFunc
```
// verifyArch verify if arch is arm64
func defaultVerifyArchFunc() bool {
	if rt.GOARCH == "arm64" {
		return true
	}
	return false
}
```

then in a existing unit test file ```pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go```, overwrite the VerifyArchFunc to be whatever you want for each test case in order to toggle the arm code paths on. something like... 

```
// turn on arm validation/mutation
webhooks.VerifyArchFunc = func() bool {return true}
```

That lets us exercise this webhook logic despite what architecture we run on.

Reply:
oh, and i also mean to mention that the `VerifyArchFunc` should probably be named something arm specific, like IsArchARMFunc, or somethign similar that indicates it's just related to arm.

Reply:
> this webhook logic all makes sense to me. Can we get a unit test to exercise this logic?
> 
> Because of the way you've gated all the arm defaults/validation behind this `verifyArch()` function, it would actually be pretty trivial to mock this arm validation/defaulting for all architectures.
> 
> at the top of the file add something like this...
> 
> ```
> var VerifyArchFunc = defaultVerifyArchFunc
> ```
> 
> then rename verifyArch function to defaultVerifyArchFunc
> 
> ```
> // verifyArch verify if arch is arm64
> func defaultVerifyArchFunc() bool {
> 	if rt.GOARCH == "arm64" {
> 		return true
> 	}
> 	return false
> }
> ```
> 
> then in a existing unit test file `pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go`, overwrite the VerifyArchFunc to be whatever you want for each test case in order to toggle the arm code paths on. something like...
> 
> ```
> // turn on arm validation/mutation
> webhooks.VerifyArchFunc = func() bool {return true}
> ```
> 
> That lets us exercise this webhook logic despite what architecture we run on.

Very good suggestion!